monitor/monitor.go

```diff
@@ -111,12 +107,58 @@ func (m *Monitor) handleConnection(server net.Listener) {
 		}
 
 		mutex.Lock()
+		first := listeners.Len() == 0
 		listeners.PushBack(conn)
 		log.WithField("count.listener", listeners.Len()).Info("New monitor connected.")
+		/* Since this is our first listener that means the event loop
+		 * is not running and has to be started.
+		 */
+		if first {
+			go m.mainEventLoop()
+		}
 		mutex.Unlock()
 	}
 }
 
+// mainEventLoop is responsible for reading events from the perf ring buffer
+// and manage the connections to the socket
+func (m *Monitor) mainEventLoop() {
+	log.Info("Starting monitoring traffic")
+	me, err := bpf.NewPerCpuEvents(eventConfig)
+	if err != nil {
+		log.WithError(err).Error("Error while starting monitor")
+		return
+	}
+	monitorEvents = me
+
+	for {
+		purgeInactiveListeners()
```

Comment:
Should we do the numListeners() == 0 check (the else on line 151 below) here? this then means we can return before the Poll if there are no listeners, and also means that if we do the Poll call we always read data in the todo > 0 case (I think this was rlenglet's concern earlier).
The general structure of the function would then look like:
``` golang
bpf.NewCpuEvents()
for {
	purgeInactiveListeners()
	if numListeners == 0 {
		CloseAll()
		return
	}
	Poll()
	if todo > 0 {
		ReadAll()
	}
}
```

An alternative would be to defer monitorEventsCloseAll() and the error handling for it and if `numListeners == 0` simply returns from the function/breaks the loop to allow the function to exit.

Reply:
Will change it to the suggested but deferring the closing of the perf event side might lead to an invalid memory address or nil pointer dereference so we can't do that.